### ISSUE TITLE
reportUnusedCallResult does not catch unused await-ed values

### ISSUE THREAD

**Issue by hcarty** [microsoft/pyright#1463]
**Describe the bug**

`reportUnusedCallResult` does not detect an unused `await`ed value.

**To Reproduce**

In vscode + pylance:

```python
# pyright: strict, reportUnusedCallResult=true

async def get_string_async() -> str:
    return "A string"

async def await_string() -> None:
    await get_string_async()  # No error here

def get_string() -> str:
    return "A string"

def call_string() -> None:
    get_string()  # Gives an error as expected
```

**Expected behavior**
I would expect both `await_string` and `call_string` to contain an error.

**VS Code extension or command-line**
Pylance 2021.2.0 with pyright 1.1.108 + unreleased changes.

Thank you again for adding this feature to pyright!  While I don't know that it's caught any issues with code already in production in our codebase at work, it has prevented a few from making their way in.  Which is exactly what I want from such a tool.

**Comment by hcarty** [comment]
After writing this, I realized that the issue is mostly likely that `await`ing a value results in a value, not a callable's result.  Given that, this may end up being more of a feature request to detect expressions which are unused and unassigned:
```python
async def get() -> str:
    return "a string"

# In this imaginary check, any of the lines in f would be an error
async def f() -> None:
    1
    "hello"
    await get()
```
If that makes more sense I'm happy to change this issue into a feature request or close this and open a separate request issue.

**Comment by erictraut** [comment]
When I implemented the `reportUnusedCallResult` setting, I considered an `await` to be a "use" of the return result. It sounds like you want to it be treated as though the `await` is effectively part of the call expression?

**Comment by hcarty** [comment]
Thank you for responding so quickly!

> It sounds like you want to it be treated as though the await is effectively part of the call expression?

I can see this working two ways.  One is what you said - treat the `await` as part of the call expression.  That would cover `await get()` in the example above.

The other is to have a separate check for unassigned and unused expressions which aren't `None`.  A potential `reportUnusedExpression` setting would catch each of `1`, `"hello"` and `await get()` in `f`.  Each are expressions which are not assigned to a variable or used in a larger expression and each are not `None`.

I would be very happy to have either of these in pyright.  The first is a change to the semantics of `reportUnusedCallResult` and the second, at least in the case of the `await get()` example, would build on top of the check provided by `reportUnusedCallResult`.

**Comment by erictraut** [comment]
As you may remember, I was reluctant to add `reportUnusedCallResult` because I thought that very few users would find it useful. That's why it's off by default even in strict mode. The ratio of "bugs found" to "false positives that need to be fixed" strikes me as extremely low. I think that ratio would be similarly low (or even lower) for general expressions. 

It's also worth nothing that call expressions are a subset of expressions, so a `reportUnusedExpression` setting would be somewhat redundant.

Since the existing `reportUnusedCallResult` feature is probably being used only by you and a handful of other developers, I'm OK modifying its behavior to suit your needs here. If you want the `await` to be considered part of a call expression, I can make that change.

**Comment by hcarty** [comment]
Thank you also for offering to change the checks behavior - Please do extend `await` to be part of the call expression.

I do remember, and am quite grateful for the addition of `reportUnusedCallResult`.  In my experience since its introduction the check has been helpful in projects, particularly when trying new libraries and acclimating new developers to an existing codebase.  I've seen it catch multiple mistakes before they can become bugs.

I do think a more general `reportUnusedExpression` would be helpful because it would cover the current `reportUnusedCallResult` cases and catch additional cases which would effectively be dead code.  I also completely understand not implementing a feature without proven worth - each new feature is a new and ongoing support burden for the project and you as a maintainer.

**Comment by erictraut** [comment]
This will be included in the next release of pyright.

**Comment by hcarty** [comment]
Thank you @erictraut!

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.110, which I just published.
